README.md

# Spark-optimizer

[![Build Status](https://api.travis-ci.org/delijati/spark-optimizer.svg?branch=master)](https://travis-ci.org/delijati/spark-optimizer)

Optimize spark settings (for cluster aka yarn run)

Original source: http://c2fo.io/c2fo/spark/aws/emr/2016/07/06/apache-spark-config-cheatsheet/

## Usage

Install:

    $ virtualenv env
    $ env/bin/pip install spark-optimizer

Dev install:

    $ virtualenv env
    $ env/bin/pip install -e .


Generate settings for `c4.4xlarge` with `4` nodes:

    $ env/bin/spark-optimizer c4.4xlarge 4
    Optimal numPartitions: 162 
    {'spark.default.parallelism': '108',
     'spark.driver.cores': '2',
     'spark.driver.maxResultSize': '3481m',
     'spark.driver.memory': '3481m',
     'spark.driver.memoryOverhead': '614m',
     'spark.executor.cores': '2',
     'spark.executor.instances': '27',
     'spark.executor.memory': '3481m',
     'spark.executor.memoryOverhead': '614m'}

Update instance info:

    $ env/bin/python spark_optimizer/emr_update.py


CHANGES.md
# CHANGES

0.1.9 (unreleased)
------------------

- Nothing changed yet.


0.1.8 (2020-02-14)
------------------

- add calculation for `numPartitions`
- dropping pypy and python3.4


0.1.7 (2019-12-09)
------------------

- set `long_description_content_type="text/markdown"`


0.1.6 (2019-12-09)
------------------

- fix docs


0.1.5 (2019-12-09)
------------------

- update ``emr_instance.yaml``


0.1.4 (2019-03-10)
------------------

- add ec2 and emr cost to yaml


0.1.3 (2019-03-08)
------------------

- add emr cost to yaml
- export load yaml file
- make ``memory_overhead_coefficient`` editable


0.1.2 (2019-02-20)
------------------

- unpin the versions
- rename cli from _ to -


0.1.1 (2018-09-12)
------------------

- fix email


0.1.0 (2018-09-12)

- initial release


test_setting.py
import unittest
from emr_setting import calculate_spark_settings


class TestSparkSettings(unittest.TestCase):
    def test_calc_spark_settings(self):
        ret, _ = calculate_spark_settings("c4.large", 4)
        self.assertDictEqual(
            ret, {
                'spark.executor.instances': "3",
                'spark.executor.memoryOverhead': "384m",
                'spark.executor.memory': "1740m",
                'spark.driver.memoryOverhead': "384m",
                'spark.driver.memory': "1740m",
                'spark.driver.maxResultSize': '1740m',
                'spark.executor.cores': "1",
                'spark.driver.cores': "1",
                'spark.default.parallelism': "6"
            })

        ret, opt = calculate_spark_settings("c4.8xlarge", 4)
        self.assertDictEqual(
            ret, {
                'spark.executor.instances': "31",
                'spark.executor.memoryOverhead': "1075m",
                'spark.executor.memory': "6092m",
                'spark.driver.memoryOverhead': "1075m",
                'spark.driver.memory': "6092m",
                'spark.driver.maxResultSize': "6092m",
                'spark.executor.cores': "4",
                'spark.driver.cores': "4",
                'spark.default.parallelism': "248"
            })


test/test_cli.py
import unittest
import pytest

from spark_optimizer.cli import main


class TestCLI(unittest.TestCase):

    @pytest.fixture(autouse=True)
    def capsys(self, capsys):
        self.capsys = capsys

    def test_cli(self):
        with pytest.raises(SystemExit) as ex:
            main([])
        self.assertEqual(ex.value.code, 1)
        out, err = self.capsys.readouterr()
        self.assertTrue("Two args are needed! (type, number nodes)" in out)
        # working
        with pytest.raises(SystemExit) as ex:
            main(["me", "c4.xlarge", "4"])
        self.assertEqual(ex.value.code, 0)
        out, err = self.capsys.readouterr()
        self.assertTrue("spark.default.parallelism" in out)


test/test_optimizer.py
import unittest

from spark_optimizer.optimizer import calculate_spark_settings


class TestSparkSettings(unittest.TestCase):
    def test_calc_spark_settings(self):
        ret, _ = calculate_spark_settings("c4.large", 4)
        self.assertDictEqual(
            ret, {
                'spark.executor.instances': "3",
                'spark.executor.memoryOverhead': "384m",
                'spark.executor.memory': "1740m",
                'spark.driver.memoryOverhead': "384m",
                'spark.driver.memory': "1740m",
                'spark.driver.maxResultSize': '1740m',
                'spark.executor.cores': "1",
                'spark.driver.cores': "1",
                'spark.default.parallelism': "6"
            })

        ret, opt = calculate_spark_settings("c4.8xlarge", 4)
        self.assertDictEqual(
            ret, {
                'spark.executor.instances': "31",
                'spark.executor.memoryOverhead': "1075m",
                'spark.executor.memory': "6092m",
                'spark.driver.memoryOverhead': "1075m",
                'spark.driver.memory': "6092m",
                'spark.driver.maxResultSize': "6092m",
                'spark.executor.cores': "4",
                'spark.driver.cores': "4",
                'spark.default.parallelism': "248"
            })


spark_optimizer/optimizer.py
import os
import math
import sys

import ruamel.yaml as yaml


def plot(max_executor, ret):
    import numpy as np
    import pandas as pd
    import matplotlib.pyplot as plt

    df = pd.DataFrame(dict(x=np.arange(1, max_executor)))
    df["unused_memory_per_node"] = np.floor(45 / df.x)
    # df = df.drop(df[<some boolean condition>].index)
    df = pd.DataFrame(ret)
    df[['unused_cores', 'unused_mem']].plot()
    plt.show()
    normalized_df = (df - df.mean()) / df.std()
    # normalized_df = (df-df.min())/(df.max()-df.min())
    normalized_df[['unused_cores', 'unused_mem']].plot()
    plt.show()
    df[['sum']].plot()
    plt.show()


def load_emr_instance():
    root = os.path.dirname(os.path.abspath(__file__))
    FILE = os.path.join(root, "emr_instance.yaml")
    with open(FILE) as f:
        ret = yaml.safe_load(f)
    return ret


def calculate_spark_settings(instance_type, num_slaves, max_executor=192,
                             memory_overhead_coefficient=0.15,
                             num_partitions_factor=3):
    """
    More info:

    http://c2fo.io/c2fo/spark/aws/emr/2016/07/06/apache-spark-config-cheatsheet/
    """
    all_instances = load_emr_instance()
    inst = all_instances[instance_type]

    executor_memory_upper_bound = 64
    executor_core_upper_bound = 5
    available_memory = inst["memory"] - 1
    available_cpu = inst["cpu"] - 1
    parallelism_per_core = 2

    ret = []

    for x in range(1, max_executor):
        total_memory_per_executor = math.floor(available_memory / x)
        unused_memory_per_node = available_memory - (
            x * total_memory_per_executor)
        total_core_per_executor = available_cpu / x
        unused_cores_per_node = available_cpu - (
            x * math.floor(total_core_per_executor))
        overhead_mem = total_memory_per_executor * memory_overhead_coefficient
        mem_executer = total_memory_per_executor - overhead_mem
        cores_executer = math.floor(available_cpu / x)

        if (total_memory_per_executor < executor_memory_upper_bound and
                total_core_per_executor < executor_core_upper_bound):
            ret.append({
                "executor_per_node": x,
                "overhead_mem": max(overhead_mem * 1024, 384),
                "unused_cores": unused_cores_per_node,
                "mem_executer": mem_executer,
                "cores_executer": cores_executer,
                "unused_mem": unused_memory_per_node,
                "sum": unused_cores_per_node + unused_memory_per_node
            })

    val, idx = min(
        (val, idx) for (idx, val) in enumerate([x["sum"] for x in ret]))

    # plot(max_executor, ret)

    opt = ret[idx]
    executer_instances = (opt["executor_per_node"] * num_slaves) - 1
    parallelism = (
        executer_instances * opt["cores_executer"] * parallelism_per_core)

    num_partitions = int(opt["cores_executer"] * executer_instances
                         * num_partitions_factor)

    print("Optimal numPartitions: %s " % num_partitions, file=sys.stderr)

    ret = {
        "spark.executor.instances": str(executer_instances),
        "spark.executor.memoryOverhead": "%sm" % int(opt["overhead_mem"]),
        "spark.executor.memory": "%sm" % int(opt["mem_executer"] * 1024),
        "spark.driver.memoryOverhead": "%sm" % int(opt["overhead_mem"]),
        "spark.driver.memory": "%sm" % int(opt["mem_executer"] * 1024),
        "spark.driver.maxResultSize": "%sm" % int(opt["mem_executer"] * 1024),
        "spark.executor.cores": str(int(opt["cores_executer"])),
        "spark.driver.cores": str(int(opt["cores_executer"])),
        "spark.default.parallelism": str(int(parallelism)),
    }

    return ret, opt


spark_optimizer/emr_update.py
import os

import requests
import collections

import ruamel.yaml

here = os.path.abspath(os.path.dirname(__file__))
FILE = os.path.join(here, "emr_instance.yaml")


def main():
    res = requests.get(
        "https://raw.githubusercontent.com/powdahound/ec2instances.info/master/www/instances.json" # noqa
    )

    emr = {}
    if os.path.exists(FILE):
        with open(FILE) as f:
            emr = ruamel.yaml.safe_load(f.read())

    inst = res.json()

    for i in inst:
        if i['emr']:
            x = i['instance_type']
            emr[x] = {}
            emr[x]["gpu"] = i["GPU"]
            emr[x]["cpu"] = i["vCPU"]
            emr[x]["memory"] = i["memory"]
            pricing = {}
            for k in i["pricing"]:
                pricing[k] = i["pricing"][k].get('emr')
            emr[x]["pricing"] = pricing

    with open(FILE, "w+") as f:
        ruamel.yaml.dump(collections.OrderedDict(sorted(emr.items())), f,
                         default_flow_style=False)


if __name__ == "__main__":
    main()


emr_setting.py
import os
import math
import sys
import yaml
import pprint as pp


def plot(max_executor, ret):
    import pandas as pd
    import numpy as np
    import matplotlib.pyplot as plt

    df = pd.DataFrame(dict(x=np.arange(1, max_executor)))
    df["unused_memory_per_node"] = np.floor(45 / df.x)
    # df = df.drop(df[<some boolean condition>].index)
    df = pd.DataFrame(ret)
    df[['unused_cores', 'unused_mem']].plot()
    plt.show()
    normalized_df = (df - df.mean()) / df.std()
    # normalized_df = (df-df.min())/(df.max()-df.min())
    normalized_df[['unused_cores', 'unused_mem']].plot()
    plt.show()
    df[['sum']].plot()
    plt.show()


def calculate_spark_settings(instance_type, num_slaves, max_executor=192):
    """
    More info:

    http://c2fo.io/c2fo/spark/aws/emr/2016/07/06/apache-spark-config-cheatsheet/
    """
    root = os.path.dirname(os.path.abspath(__file__))
    FILE = os.path.join(root, "emr_instance.yaml")
    with open(FILE) as f:
        all_instances = yaml.load(f)
    inst = all_instances[instance_type]

    memory_overhead_coefficient = 0.15
    executor_memory_upper_bound = 64
    executor_core_upper_bound = 5
    available_memory = inst["memory"] - 1
    available_cpu = inst["cpu"] - 1
    parallelism_per_core = 2

    ret = []

    for x in range(1, max_executor):
        total_memory_per_executor = math.floor(available_memory / x)
        unused_memory_per_node = available_memory - (
            x * total_memory_per_executor)
        total_core_per_executor = available_cpu / x
        unused_cores_per_node = available_cpu - (
            x * math.floor(total_core_per_executor))
        overhead_mem = total_memory_per_executor * memory_overhead_coefficient
        mem_executer = total_memory_per_executor - overhead_mem
        cores_executer = math.floor(available_cpu / x)

        if (total_memory_per_executor < executor_memory_upper_bound and
                total_core_per_executor < executor_core_upper_bound):
            ret.append({
                "executor_per_node": x,
                "overhead_mem": max(overhead_mem * 1024, 384),
                "unused_cores": unused_cores_per_node,
                "mem_executer": mem_executer,
                "cores_executer": cores_executer,
                "unused_mem": unused_memory_per_node,
                "sum": unused_cores_per_node + unused_memory_per_node
            })

    val, idx = min(
        (val, idx) for (idx, val) in enumerate([x["sum"] for x in ret]))

    # plot(max_executor, ret)

    opt = ret[idx]
    executer_instances = (opt["executor_per_node"] * num_slaves) - 1
    parallelism = (
        executer_instances * opt["cores_executer"] * parallelism_per_core)

    ret = {
        "spark.executor.instances": str(executer_instances),
        "spark.executor.memoryOverhead": "%sm" % int(opt["overhead_mem"]),
        "spark.executor.memory": "%sm" % int(opt["mem_executer"] * 1024),
        "spark.driver.memoryOverhead": "%sm" % int(opt["overhead_mem"]),
        "spark.driver.memory": "%sm" % int(opt["mem_executer"] * 1024),
        "spark.driver.maxResultSize": "%sm" % int(opt["mem_executer"] * 1024),
        "spark.executor.cores": str(int(opt["cores_executer"])),
        "spark.driver.cores": str(int(opt["cores_executer"])),
        "spark.default.parallelism": str(int(parallelism)),
    }

    return ret, opt


def main(argv):
    if len(argv) < 3 or len(argv) > 3:
        raise Exception("Two args are needed! (type, number nodes)")
    pp.pprint(calculate_spark_settings(argv[1], int(argv[2]))[0])


if __name__ == "__main__":
    main(sys.argv)
